bin/varnishd/http2/cache_http2_session.c

```diff
@@ -444,17 +445,59 @@ h2_new_session(struct worker *wrk, void *arg)
 static void
 h2_waitlist(struct worker *wrk, struct req *req)
 {
+	struct objhead *oh;
+	struct h2_req *r2;
 
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(req, REQ_MAGIC);
+	oh = req->hash_objhead;
+	CHECK_OBJ_NOTNULL(oh, OBJHEAD_MAGIC);
+	CAST_OBJ_NOTNULL(r2, req->top->topreq->transport_priv, H2_REQ_MAGIC);
+
+	if (DO_DEBUG(DBG_WAITINGLIST))
+		VSLb(req->vsl, SLT_Debug, "on h2 waiting list <%p>", r2);
+
+	Lck_AssertHeld(&oh->mtx);
+	AN(req->waitinglist);
+	AZ(req->wrk);
+
+	assert(oh->refcnt > 0);
+	oh->refcnt++;
+
+	Lck_Lock(&r2->h2sess->sess->mtx);
+	AZ(req->transport_objhead);
+	req->transport_objhead = oh;
```

Comment:
The variable `req->transport_objhead` should be renamed `req->toptransport_objhead`, to reflect that it is "owned" by the top transport. As it is it would be trouble if e.g. esi_deliver wanted to use it. Same argument goes for `t_list`.

Reply:
Or, maybe we could move `req::t_list` and `req::transport_objhead` to `reqtop::w_list` and `reqtop::hash_objhead`?